agent/lib/kontena/workers/node_info_worker.rb

```diff
@@ -236,6 +253,58 @@ def calculate_container_time(container)
       0
     end
 
+    # @param [Array<Vmstat::Cpu>] prev_cpus
+    # @param [Array<Vmstat::Cpu>] current_cpu
+    # @return [Hash] { :num_cores, :system, :user, :idle }
+    def calculate_cpu_usage(prev_cpus, current_cpus)
+      result = {
+        num_cores: prev_cpus.size,
+        system: 0.0,
+        user: 0.0,
+        idle: 0.0
+      }
+
+      prev_cpus.zip(current_cpus).map { |prev_cpu, current_cpu|
+        system_ticks = current_cpu.system - prev_cpu.system
+        user_ticks = current_cpu.user - prev_cpu.user
+        idle_ticks = current_cpu.idle - prev_cpu.idle
+
+        total_ticks = (system_ticks + user_ticks + idle_ticks).to_f
+
+        {
+          system: (system_ticks / total_ticks) * 100.0,
+          user: (user_ticks / total_ticks) * 100.0,
+          idle: (idle_ticks / total_ticks) * 100.0
+        }
+      }.inject(result) { |memo, cpu_core|
+        memo[:system] += cpu_core[:system]
+        memo[:user] += cpu_core[:user]
+        memo[:idle] += cpu_core[:idle]
+        memo
+      }
+    end
+
+    # @param [Vmstat::NewtworkInterface] previous_iface
+    # @param [Vmstat::NetworkInterface] current_iface
+    # @param [Number] interval
+    # @return [Hash]
+    def calculate_network_traffic(previous_iface, current_iface, interval)
+      in_bytes_per_second = (current_iface.in_bytes - previous_iface.in_bytes) / interval
+      out_bytes_per_second = (current_iface.out_bytes - previous_iface.out_bytes) / interval
+
+      {
+        interface_name: current_iface.name,
+        in_bytes_per_second: in_bytes_per_second,
+        out_bytes_per_second: out_bytes_per_second
+      }
+    end
+
+    def get_network_interface
+      Vmstat.network_interfaces.select { |x| x.ethernet? and x.in_bytes > 0 and x.out_bytes > 0 }
```

Comment:
Selecting the network interface with the most traffic isn't a good idea, each node will have several network interfaces with different levels of traffic. Enumerating the interfaces isn't that straightforward either, since they will vary across different platforms....

On virtual machines, it seems like `ethernet?` may exclude the bridge and veth interfaces, leaving only the virtualized ethernet interfaces: `Vmstat.network_interfaces.select{|x|x.ethernet?}.map{|x| x.name} => [:eth0, :eth1]` Using Kontena terms, `eth1` would be the `private_interface` (`ENV['KONTENA_PEER_INTERFACE']`), and then the remaining `eth0` could be assumed to be the `public_interface`. However, on some platforms like Packet, the interfaces will be entirely different... 

Other interesting interfaces will well-known names would be the `docker0` bridge interface, and the `weave` bridge interface. The `docker0` bridge interface stats are the total of all the container `eth0` network stats, whereas the `weave` bridge interface stats would be the total of all container `ethwe` network stats.